File: LeetCode/48. Rotate Image/48.java
class Solution {
    public void rotate(int[][] matrix) {
        int n = matrix.length;

        // Transpose the matrix
        for (int i = 0; i < n; i++) {
            for (int j = i; j < n; j++) {
                int temp = matrix[i][j];
                matrix[i][j] = matrix[j][i];
                matrix[j][i] = temp;
            }
        }

        // Reverse each row
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n / 2; j++) {
                int temp = matrix[i][j];
                matrix[i][j] = matrix[i][n - 1 - j];
                matrix[i][n - 1 - j] = temp;
            }
        }
    }
}


/*
 class Solution {
    public void rotate(int[][] matrix) {
        int k=matrix.length;
        int temp;
        for(int i=0;i<k;i++){
            for(int j=i;j<k;j++){
                temp=matrix[i][j];
                matrix[i][j]=matrix[j][i];
                matrix[j][i]=temp;
                
            }
        }
        for(int x=0;x<matrix.length;x++){
            int left=0;
            int right=matrix.length-1;
            while(left<right){
                temp=matrix[x][right];
                matrix[x][right]=matrix[x][left];
                matrix[x][left]=temp;
                left++;
                right--;
                
            }
        }

    }
}
 */
/*
 class Solution {
    public void rotate(int[][] matrix) {
        int l = matrix.length;
        int i=0;
        int j= l-1;
        while(i<j) {
            for(int count=0; i+count<j; count++){
                rotate(matrix, i, j, count);
            }
            i++; j--;
        }
    }

    public void rotate(int[][] matrix, int i, int j, int count ) {
        int temp = matrix[i][i+count];
        matrix[i][i+count] = matrix[j-count][i];
        matrix[j-count][i] = matrix[j][j-count];
        matrix[j][j-count] = matrix[i+count][j];
        matrix[i+count][j] = temp;
    }
}
 */

 /*
  class Solution {
    private void transpose(int[][] matrix) {
        for(int i=0; i<matrix.length; i++) {
            for(int j=i; j<matrix[0].length; j++) {
                int temp = matrix[i][j];
                matrix[i][j] = matrix[j][i];
                matrix[j][i] = temp;
            }
        }
    }
    private void reverseRows(int[][] matrix) {
        for(int r=0; r<matrix.length; r++) {
            int left = 0;
            int right = matrix.length-1;

            while(left < right) {
                int temp = matrix[r][left];
                matrix[r][left] = matrix[r][right];
                matrix[r][right] = temp;

                left++;
                right--;
            }
        }

    }
    public void rotate(int[][] matrix) {
        transpose(matrix);
        reverseRows(matrix);
    }
}
  */

File: LeetCode/48. Rotate Image/48.py
class Solution(object):
    def rotate(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: None Do not return anything, modify matrix in-place instead.
        """

        left =0
        right =len(matrix) -1

        while left < right:
            for i in range(right-left):
                top = left
                bottom = right

                holder = matrix[top][left+i]

                #swap bl tl br-bl 
                matrix[top][left+i] = matrix[bottom-i][left]

                matrix[bottom-i][left] = matrix[bottom][right-i]

                matrix[bottom][right-i] = matrix[top+i][right]

                matrix[top+i][right] = holder

            left +=1
            right -=1
            

'''
class Solution(object):
    def rotate(self, matrix):
        matrix[:] = zip(*matrix[::-1])
        
'''

'''
class Solution(object):
    def rotate(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: None Do not return anything, modify matrix in-place instead.
        """
        n = len(matrix)
        
        # Transpose the matrix
        for i in range(n):
            for j in range(i, n):
                matrix[i][j], matrix[j][i] = matrix[j][i], matrix[i][j]
        
        # Reverse each row
        for i in range(n):
            matrix[i].reverse()
'''